### Practice: Large Language Models and Their Implications
<!-- ![img](https://substackcdn.com/image/fetch/f_auto,q_auto:good,fl_progressive:steep/https%3A%2F%2Fbucketeer-e05bbc84-baa3-437e-9518-adb32be77984.s3.amazonaws.com%2Fpublic%2Fimages%2F4470ce74-e595-4750-92a5-5f21f040df6d_577x432.jpeg) -->
![img](https://i.imgur.com/QGYa2J8.jpeg)

In this notebook, you're gonna play with some of the largest language models on the Internet.

_Based on works of: Tim Dettmers, Ruslan Svirschevsky, Artem Chumachenko, Younes Belkada, Felix Marty, Yulian Gilyazev, Gosha Zolotov, Andrey Ishutin,  Elena Volf, Artemiy Vishnyakov, Svetlana Shirokovskih.

### Part 1: prompt engineering (4 points total)

In the assignment, we'll use public APIs that host the 100B+ models for inference. Your task is to prompt-engineer the model into solving a few tasks for you.


__Which API?__ You are free to use any publicly available API for general LM -- as long as it's __not a chat assistant__. So, gpt 3.5 is fine, but chatGPT is not. Here's a few options:

- BLOOM API - [bigscience/bloom](https://huggingface.co/bigscience/bloom) (on the right; recommended)
- OpenAI API (via VPN) - [openai.com/api](https://openai.com/api/)
- AI21 Jurrasic API - [ai21.com](https://www.ai21.com/blog/announcing-ai21-studio-and-jurassic-1)

These APIs may require you to create a (free) account on their platform. Please note that some APIs also have paid subscriptions. __You do not need to pay them__, this assignment was designed to be solved using free-tier subscriptions. If no APIs work for you, you can also solve these tasks with the 6.7B model that you will find later in this notebook - but this will make the tasks somewhat harder.

__Quests:__ you will need to solve 4 problems. For each one, please attach a short __description__ of your solution and a __screenshot__ from the API you use. _[If you use python APIs, show your python code with outputs]_

__Example:__ Tony is talking to Darth Vader ([BLOOM API](https://huggingface.co/bigscience/bloom)). Black text is written manually, blue text is generated.
<hr>

![img](https://i.imgur.com/a1QhKF7.png)
<hr>

__It is fine to roll back a few times,__ e.g. in the example above, the model first generated Vader lines twice in a row, and we rolled that back. However, if you need more than 1-2 rollbacks per session, you should probably try a different prompt.

__Task 1 (1 pt):__ arange a conversation between any two of the following:

- a celebrity or politician of your choice
- any fictional character (except Darth Vader)
- yourself

Compare two setups: a) you prompt with character names only b) you supply additional information (see example).

In [ ]:
!pip uninstall -y huggingface_hub
!pip install -U huggingface_hub

from huggingface_hub import login

login()

In [ ]:
def predict_response(model, prompt, max_tokens, temp):
  client = InferenceClient(provider="novita", token=True)

  response = client.chat_completion(
      model="deepseek-ai/DeepSeek-V3-0324",
      messages=[{"role": "user", "content": prompt}],
      max_tokens=max_tokens,
      temperature=temp,
  )

  return response.choices[0].message.content


In [ ]:
from huggingface_hub import InferenceClient

model="deepseek-ai/DeepSeek-V3-0324"

prompt = """
Write a dialogue
Winston Churchill is talking to Joseph Stalin.

Winston: Hello, Joseph. Do you think rockets could use magic?
Joseph:
"""
predict_response(model, prompt, 200, 0.8)

**Dialogue:**  

**Winston:** Hello, Joseph. Do you think rockets could use magic?  

**Joseph:** *[chuckles, puffing on his pipe]* Magic? Winston, rockets are built by engineers, not wizards. The only "magic" we need is the will of the proletariat and the power of Soviet science.  

**Winston:** Ah, but consider—what if a rocket could fly not by fuel, but by sheer British ingenuity and a dash of sorcery?  

**Joseph:** *[grinning darkly]* If magic worked, Churchill, I would have turned capitalists into frogs by now. Stick to whiskey and speeches. Rockets need steel, not spells.  

**Winston:** A shame. I rather fancied a spellbound V-2. Well, I suppose we’ll have to settle for boring old physics.  

**Joseph:** *[nodding]* Da. Physics wins wars. Magic is for fairy tales


In [ ]:
prompt = """
Write a dialogue and stay within the token limit
Winston Churchill is the Prime Minister of the United Kingdom during World War II.
He is witty, eloquent, confident, and likes long rhetorical speeches.

Joseph Stalin is the leader of the Soviet Union.
He is cold, pragmatic, suspicious, and speaks in short, harsh statements focused on power and war.

Winston Churchill is talking to Joseph Stalin about rockets and magic.

Winston: Hello, Joseph. Do you think rockets could use magic?
Joseph:

"""

predict_response(model, prompt, 200, 0.8)


**Joseph:** Rockets need steel, fuel, discipline. Magic is bourgeois fantasy.  

**Winston:** My dear fellow, even steel obeys the laws of physics—why not the unseen forces of the universe? Imagine a V-2 rocket, but with a dash of Merlin’s spark!  

**Joseph:** Hmph. If magic won shells, we’d have wizards at Stalingrad. We had rifles. We won.  

**Winston:** Ah, but a wizard might’ve saved us both the trouble! Still, I admire your practicality. Suppose magic *were* real—would you harness it, or purge it as a threat?  

**Joseph:** If it serves the state, we use it. If not—*gestures vaguely to the gulag*—waste of time.  

**Winston:** *chuckles* Ever the pragmatist. Very well—let’s stick to rockets. And vodka.  

**Joseph:**


Conclusion:

The second prompt produced a noticeably better dialogue. By adding descriptions of the characters, the model was able to maintain more consistent personalities and speaking styles throughout the conversation. Churchill sounded more eloquent and rhetorical, while Stalin appeared more cold, direct, and pragmatic.  

Compared to the first version, the dialogue in part B was more coherent and believable. The additional context helped the model better understand the historical and personality differences between the two characters, which improved the overall quality of the generation.

__Please choose task 2a or 2b (1pt)__ depending on your model (you can do both, but you will be awarded points for one of these two tasks).

__Task 2a: (for BLOOM or other multilingual model)__ zero-shot translation. Take the first verse of [Edgar Allan Poe's "Raven"](https://www.poetryfoundation.org/poems/48860/the-raven) and __translate it into French.__ (You are free to use any other text of at least the same size)

Original text: ```
Once upon a midnight dreary, while I pondered, weak and weary,
Over many a quaint and curious volume of forgotten lore—
    While I nodded, nearly napping, suddenly there came a tapping,
As of some one gently rapping, rapping at my chamber door.
“’Tis some visitor,” I muttered, “tapping at my chamber door—
            Only this and nothing more.”
```

Verify your translation by converting french back into english using a public machine translation service.

__Task 2b: (non-BLOOM):__ toxicity classification for [SetFit/toxic_conversations](https://huggingface.co/datasets/SetFit/toxic_conversations). Make the model solve binary classification (toxic vs not toxic) in the few shot mode. For few-shot examples, use 2-3 toxic and 2-3 non-toxic non-toxic examples. Measure accuracy on at least 25 samples. You may need to try several different prompts before you find the one that works.

In [ ]:
# <your code OR writeup with screenshots>

text = """
Once upon a midnight dreary, while I pondered, weak and weary,
Over many a quaint and curious volume of forgotten lore—
    While I nodded, nearly napping, suddenly there came a tapping,
As of some one gently rapping, rapping at my chamber door.
“’Tis some visitor,” I muttered, “tapping at my chamber door—
            Only this and nothing more.”
"""

prompt = f"""
translate it into French: {text}
"""

predict_response(model, prompt, 500, 0)

Voici la traduction en français :

« Une fois, par un minuit lugubre, tandis que je méditais, faible et fatigué,  
Sur maint volume curieux et bizarre de savoir oublié —  
Tandis que je donnais de la tête, presque assoupi, soudain il vint un frappement,  
Comme de quelqu’un frappant doucement, frappant à la porte de ma chambre.  
« C’est quelque visiteur, » murmurai-je, Qui frappe à la porte de ma chambre.  
— Cela et rien de plus. »

Cette traduction s'inspire de la version classique de Baudelaire, qui a adapté *The Raven* d'Edgar Allan Poe sous le titre *Le Corbeau*. J'ai ajusté certains termes pour rester fidèle au texte original tout en conservant le rythme et l'atmosphère poétique.  

(Note : Pour une référence exacte à la traduction de Baudelaire, voici son premier quatrain : *« Une fois, par un minuit lugubre, tandis que je rêvais, faible et fatigué… »*).


In [ ]:
pip install sacrebleu

In [ ]:
from sacrebleu import sentence_bleu

original = """
Once upon a midnight dreary, while I pondered, weak and weary,
Over many a quaint and curious volume of forgotten lore—
    While I nodded, nearly napping, suddenly there came a tapping,
As of some one gently rapping, rapping at my chamber door.
“’Tis some visitor,” I muttered, “tapping at my chamber door—
            Only this and nothing more.
"""

translate = """
Once, on a gloomy midnight, while I was pondering, weak and weary,
Over many a curious and bizarre volume of forgotten knowledge—
As I nodded off, almost asleep, suddenly there came a knock,
As if someone were knocking softly, knocking at my bedroom door.
“It is some visitor,” I murmured, “knocking at my chamber door.
— That and nothing more.”
"""


print(sentence_bleu(translate, [original]).score)

32.12703088491292


Conclusion:

To evaluate the quality of the translation, I translated the French version back into English and compared it with the original text using the BLEU metric. The resulting BLEU score was 32.13, which indicates a good level of similarity between the original text and the back-translation. Most of the meaning, atmosphere, and key details of the poem were preserved, although some poetic expressions and stylistic nuances were simplified during translation.


__Task 3 (1pt):__ create a prompt and few-shot examples tha make the model __change the gender pronouns__ of the main actor in a given sentence in any direction of your choice. E.g. the doctor took off _his_ mask <-> the doctor took of _her_ mask.


In [ ]:
# <your code OR writeup with screenshots>

examples = """
Input: The doctor took off his mask.
Output: The doctor took off her mask.

Input: The teacher said that he would check the homework tomorrow.
Output: The teacher said that she would check the homework tomorrow.

Input: The engineer finished his project and presented it to his manager.
Output: The engineer finished her project and presented it to her manager.
"""

prompt = f"""
change the gender pronouns

examples: {examples}

Input: The lawyer brought his documents because he expected questions from the judge.
Output:

Input: The programmer updated his code after he discovered a bug in his application.
Output:

Input: The professor reviewed his notes before giving his lecture.
Output:
"""

predict_response(model, prompt, 200, 0)

Here are the outputs with the gender pronouns changed:  

1. **Input:** The lawyer brought his documents because he expected questions from the judge.  
   **Output:** The lawyer brought her documents because she expected questions from the judge.  

2. **Input:** The programmer updated his code after he discovered a bug in his application.  
   **Output:** The programmer updated her code after she discovered a bug in her application.  

3. **Input:** The professor reviewed his notes before giving his lecture.  
   **Output:** The professor reviewed her notes before giving her lecture.  

Let me know if you'd like any further modifications!


__Task 4 (1pt):__ write a prompt and supply examples such that the model would __convert imperial units to metric units__ (miles -> kilometers; mph -> kph). More specifically, the model should rewrite a given sentence and replace all imperial units with their metric equivalents. After it works with basic distances and speed, try to find complicated examples where it does *not* work.

Please note that 1 mile is not equal to 1 km :)

In [ ]:
# <your code OR writeup with screenshots>

examples = """
Input: The store is 1 mile away.
Output: The store is 1.6 kilometers away.

Input: The car was traveling at 1 mph.
Output: The car was traveling at 1.6 kph.

Input: The runner completed 2 miles.
Output: The runner completed 3.2 kilometers.

Input: The cyclist was riding at 2 mph.
Output: The cyclist was riding at 3.2 kph.

Input: The village is 5 miles from here.
Output: The village is 8.0 kilometers from here.

Input: The train was moving at 5 mph.
Output: The train was moving at 8.0 kph.

Input: We traveled 12 miles at 12 mph.
Output: We traveled 19.3 kilometers at 19.3 kph.

Input: The airport is 20 miles away and the bus travels at 20 mph.
Output: The airport is 32.2 kilometers away and the bus travels at 32.2 kph.
"""

prompt = f"""
convert imperial units to metric units

examples: {examples}

Input: The airport is 45 miles away and we are driving at 65 mph.
Output:

Input: During the trip we traveled 120 miles in 2 hours at an average speed of 60 mph.
Output:
"""

predict_response(model, prompt, 200, 0)

Output: The airport is 72.4 kilometers away and we are driving at 104.6 kph.  

Output: During the trip we traveled 193.1 kilometers in 2 hours at an average speed of 96.6 kph.


### Part 2: local inference

Now, let's try and load the strongest model that can fit a typical Colab GPU (T4 with 16 GB as of spring 2023).

Our best candidates are the smaller versions of the best performing open source models:
- 7 Bn parameters version of [LLaMA](https://arxiv.org/pdf/2302.13971.pdf) - best for spring 2023, released by Facebook
- 7 Bn parameters version of [Falcon](https://falconllm.tii.ae) - close competitor to Llama, released in May 2023 by [Technology Innovation Institute of UAE](https://www.tii.ae).
- 6.7 Bn parameters version of [OPT](https://arxiv.org/abs/2205.01068) - top choice in this nomination in 2022, released by Facebook.

Beware: while these models are smaller than the ones in API, they're still over 60x larger than the BERT we played with last time. The code below will *just barely* fit into memory, so make sure you don't have anything else loaded. Sometimes you may need to restart runtime for the code to work.

It's a good time to restart your kernel and switch to GPU! (Runtime -> Change runtime type)
<center><img src="https://i.imgur.com/OOfDYzJ.png" width=240px></center>

In [ ]:
!pip uninstall -y transformers tokenizers accelerate
!pip install -U transformers tokenizers accelerate

In [ ]:
import torch
device = "cuda" if torch.cuda.is_available() else "cpu"
print(device)

cuda


In [ ]:
from transformers import AutoTokenizer, AutoModelForCausalLM, pipeline

model_name = "Qwen/Qwen2.5-1.5B-Instruct"

tokenizer = AutoTokenizer.from_pretrained(model_name)

model = AutoModelForCausalLM.from_pretrained(
    model_name,
    torch_dtype=torch.float16,
    device_map="auto"
)

model.to(device)

generator = pipeline(
    "text-generation",
    model=model,
    tokenizer=tokenizer
)

prompt = "Winston Churchill and Joseph Stalin are discussing rockets.\nChurchill:"
out = generator(
    prompt,
    max_new_tokens=120,
    do_sample=True,
    temperature=0.8,
    top_p=0.9,
    pad_token_id=tokenizer.eos_token_id
)

print(out[0]["generated_text"])

[transformers] `torch_dtype` is deprecated! Use `dtype` instead!


model.safetensors:   0%|          | 0.00/3.09G [00:00<?, ?B/s]

Loading weights:   0%|          | 0/338 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/242 [00:00<?, ?B/s]

[transformers] Passing `generation_config` together with generation-related arguments=({'temperature', 'max_new_tokens', 'top_p', 'pad_token_id', 'do_sample'}) is deprecated and will be removed in future versions. Please pass either a `generation_config` object OR all generation parameters explicitly, but not both.
[transformers] Both `max_new_tokens` (=120) and `max_length`(=20) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)
[transformers] Ignoring clean_up_tokenization_spaces=True for BPE tokenizer Qwen2Tokenizer. The clean_up_tokenization post-processing step is designed for WordPiece tokenizers and is destructive for BPE (it strips spaces before punctuation). Set clean_up_tokenization_spaces=False to suppress this warning, or set clean_up_tokenization_spaces_for_bpe_even_though_it_will_corrupt_output=True to force cleanup anyway.


Winston Churchill and Joseph Stalin are discussing rockets.
Churchill: "You know, I've been fascinated by the advancements in rocket technology. It's truly incredible how much progress has been made since my time."
Stalin: "Indeed, Mr. Churchill. We have seen significant developments that could revolutionize warfare and transportation."
Churchill: "I'm particularly interested in your work with the Soviet Union on this field. Can you share any details about the advancements you've achieved?"
Stalin: "We have developed a number of innovative rockets capable of delivering precision-guided payloads to specific targets. Our scientists have also advanced our understanding of aerodynamics and materials science,


## Text generation

**Comparison of strategies for language model text generation:**

| Strategy | Description | Pros & Cons |
| --- | --- | --- |
| Greedy Search | Chooses the word with the highest probability as the next word in the sequence. | **Pros:** Simple and fast. <br> **Cons:** Can lead to repetitive and incoherent text. |
| Sampling with Temperature | Introduces randomness in the word selection. A higher temperature leads to more randomness. | **Pros:** Allows exploration and diverse output. <br> **Cons:** Higher temperatures can lead to nonsensical outputs. |
| Nucleus Sampling (Top-p Sampling) | Selects the next word from a truncated vocabulary, the "nucleus" of words that have a cumulative probability exceeding a pre-specified threshold (p). | **Pros:** Balances diversity and quality. <br> **Cons:** Setting an optimal 'p' can be tricky. |
| Beam Search | Explores multiple hypotheses (sequences of words) at each step, and keeps the 'k' most likely, where 'k' is the beam width. | **Pros:** Produces more reliable results than greedy search. <br> **Cons:** Can lack diversity and lead to generic responses. |
| Top-k Sampling | Randomly selects the next word from the top 'k' words with the highest probabilities. | **Pros:** Introduces randomness, increasing output diversity. <br> **Cons:** Random selection can sometimes lead to less coherent outputs. |
| Length Normalization | Prevents the model from favoring shorter sequences by dividing the log probabilities by the sequence length raised to some power. | **Pros:** Makes longer and potentially more informative sequences more likely. <br> **Cons:** Tuning the normalization factor can be difficult. |
| Stochastic Beam Search | Introduces randomness into the selection process of the 'k' hypotheses in beam search. | **Pros:** Increases diversity in the generated text. <br> **Cons:** The trade-off between diversity and quality can be tricky to manage. |
| Decoding with Minimum Bayes Risk (MBR) | Chooses the hypothesis (out of many) that minimizes expected loss under a loss function. | **Pros:** Optimizes the output according to a specific loss function. <br> **Cons:** Computationally more complex and requires a good loss function. |

Documentation references:
- [reference for `AutoModelForCausalLM.generate()`](https://huggingface.co/docs/transformers/v4.29.1/en/main_classes/text_generation#transformers.GenerationMixin.generate)
- [reference for `AutoTokenizer.decode()`](https://huggingface.co/docs/transformers/main_classes/tokenizer#transformers.PreTrainedTokenizer.decode)
- Huggingface [docs on generation strategies](https://huggingface.co/docs/transformers/generation_strategies)

### Generation with HuggingFace

In [ ]:
prompt = 'The first discovered martian lifeform looks like'
batch = tokenizer(prompt, return_tensors='pt', return_token_type_ids=False).to(device)
print("Input batch (encoded):", batch)

output_tokens = model.generate(**batch, max_new_tokens=64, do_sample=True, temperature=0.8)
# greedy inference:                                        do_sample=False)
# beam search for highest probability:                     num_beams=4)

print("\nOutput:", tokenizer.decode(output_tokens[0].cpu()))

Input batch (encoded): {'input_ids': tensor([[  785,  1156, 11105, 48195,  1103,  2272,   627,  5868,  1075]],
       device='cuda:0'), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1]], device='cuda:0')}

Output: The first discovered martian lifeform looks like this. Credit: NASA/JPL-Caltech/MSSS
A new study has found evidence of the earliest known form of life on Mars, a single-celled organism called Glastonbury that lived approximately 3.45 billion years ago.
The findings were published in the journal Astrobiology on Wednesday.
G


In [ ]:
prompt = 'The first discovered martian lifeform looks like'
batch = tokenizer(prompt, return_tensors='pt', return_token_type_ids=False).to(device)
print("Input batch (encoded):", batch)

output_tokens = model.generate(**batch, max_new_tokens=64, do_sample=False, temperature=0.8)
# beam search for highest probability:                     num_beams=4)

print("\nOutput:", tokenizer.decode(output_tokens[0].cpu()))

[transformers] The following generation flags are not valid and may be ignored: ['temperature']. Set `TRANSFORMERS_VERBOSITY=info` for more details.


Input batch (encoded): {'input_ids': tensor([[  785,  1156, 11105, 48195,  1103,  2272,   627,  5868,  1075]],
       device='cuda:0'), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1]], device='cuda:0')}

Output: The first discovered martian lifeform looks like a tiny, red-colored insect. The discovery was made by NASA's Curiosity rover on Mars.
Curiosity found the fossilized remains of an ancient organism in sedimentary rock near its landing site at Gale Crater. It is believed to be about 3 billion years old and lived around 3.5 billion years


In [ ]:
prompt = 'The first discovered martian lifeform looks like'
batch = tokenizer(prompt, return_tensors='pt', return_token_type_ids=False).to(device)
print("Input batch (encoded):", batch)

output_tokens = model.generate(**batch, max_new_tokens=64, do_sample=True, temperature=0.8, num_beams=4)

print("\nOutput:", tokenizer.decode(output_tokens[0].cpu()))

Input batch (encoded): {'input_ids': tensor([[  785,  1156, 11105, 48195,  1103,  2272,   627,  5868,  1075]],
       device='cuda:0'), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1]], device='cuda:0')}

Output: The first discovered martian lifeform looks like this:

[![enter image description here](https://i.stack.imgur.com/6QZzR.jpg)](https://i.stack.imgur.com/6QZzR.jpg)

It's a picture of a Martian meteorite called ALH84001, which was found in Antarctica in 1


#### Low-level code for text generation

In [ ]:
prompt = "Moscow is the capital of"
# prompt = "Skippy, a young android, likes to dream about electric"

print(prompt, '\n')

voc = tokenizer.get_vocab()
voc_rev = {v:k for k, v in voc.items()}  # reverse vocab for decode

for i in range(10):
    inputs = tokenizer(prompt, return_tensors='pt', return_token_type_ids=False).to(device)
    logits = model.forward(**inputs).logits[0, -1, :]
    probs = torch.nn.functional.softmax(logits, dim=-1)
    next_token_id = torch.multinomial(probs.flatten(), num_samples=1)

    next_token = tokenizer.decode(next_token_id)
    prompt += next_token

    sorted_probs, sorted_indices = torch.sort(probs, descending=True)
    top_tokens = sorted_indices[:5]
    print(f"Step #{i} candidates:")
    for t, p in zip (top_tokens, sorted_probs):
        t = voc_rev[t.item()]
        print(f"{t:<10}: {p:.4f} ")

    print(f'\nChosen token: {next_token}', end='\n\n', flush=True)

Moscow is the capital of 

Step #0 candidates:
ĠRussia   : 0.5884 
Ġthe      : 0.1823 
Ġwhich    : 0.0394 
Ġ______   : 0.0267 
Ġ__       : 0.0247 

Chosen token:  Russia

Step #1 candidates:
.         : 0.3716 
,         : 0.2437 
Ġand      : 0.1783 
.Ċ        : 0.0755 
.ĊĊ       : 0.0287 

Chosen token:  and

Step #2 candidates:
Ġthe      : 0.3545 
Ġis       : 0.1284 
Ġone      : 0.0939 
Ġits      : 0.0698 
Ġa        : 0.0597 

Chosen token:  the

Step #3 candidates:
Ġlargest  : 0.6831 
Ġmost     : 0.0495 
Ġsecond   : 0.0465 
Ġcountry  : 0.0310 
ĠRussian  : 0.0230 

Chosen token:  largest

Step #4 candidates:
Ġcity     : 0.9663 
ĠRussian  : 0.0069 
Ġand      : 0.0060 
Ġurban    : 0.0043 
Ġmet      : 0.0041 

Chosen token:  city

Step #5 candidates:
Ġin       : 0.8657 
Ġof       : 0.0399 
.         : 0.0194 
Ġby       : 0.0191 
,         : 0.0107 

Chosen token:  in

Step #6 candidates:
ĠEurope   : 0.4568 
Ġthe      : 0.3394 
ĠRussia   : 0.1309 
Ġboth     : 0.0201 
ĠEuras    : 0.0090 


**Task 5: write code for nucleus sampling generation (2 points)**:

Use the `nucleus_sampling()` template below. Look at the detailed generation code above for inspiration. __Please do not use model.generate__.

**Bonus task: write code for beam search (3 bonus points)**

In [ ]:
from typing import Tuple, List
def nucleus_sampling(model, tokenizer, prompt: str, prob: float = 0.5) -> Tuple[str, List[str]]:
    """generates the next token from the nucleus of tokens with cumulative probability up to param:prob"""
    inputs = tokenizer(prompt, return_tensors='pt', return_token_type_ids=False).to(device)
    logits = model.forward(**inputs).logits[0, -1, :]
    probs = torch.nn.functional.softmax(logits, dim=-1)

    sorted_probs = probs.sort(descending=True) # отсортируем вероятности от большего к меньшему
    nucleus_cumsum = torch.cumsum(sorted_probs.values, dim=-1) # посчитаем кумулятивную сумму

    mask = nucleus_cumsum <= prob # сделаем маску из токенов, которе проходят порог вероятности
    mask = torch.roll(mask, shifts=1) # нужно включить первый токен, который пересёк top_p, иначе сумма ядра может быть меньше top_p
    mask[0] = True

    nucleus_probs = sorted_probs.values[mask] # сделаем ядро токенов
    nucleus_ids = sorted_probs.indices[mask] # сделаем ядро idx
    nucleus_probs = nucleus_probs / nucleus_probs.sum() # отнормируем вероятности

    sampled_position = torch.multinomial(nucleus_probs, num_samples=1)
    sampled_token_id = nucleus_ids[sampled_position] # выберем токен с учетом вероятностей из ядра
    sampled_token = tokenizer.decode(sampled_token_id).strip()
    possible_tokens = [tokenizer.decode(token_id).strip() for token_id in nucleus_ids]

    # sampled_token should be a string token that was generated
    # possible_tokens should be a list of all tokens that have non-zero probability
    return sampled_token, possible_tokens

In [ ]:
# Tests for nucleus sampling
test_prompt = "Elbrus is the highest"
next_token, possible_tokens = nucleus_sampling(model, tokenizer, test_prompt, prob=0.25)
print(test_prompt, next_token, possible_tokens)
assert next_token in possible_tokens
assert 3 <= len(possible_tokens) <= 3
assert sorted(possible_tokens) == ['mountain', 'peak', 'point']

test_prompt = "Large language models can learn to"
next_token, possible_tokens = nucleus_sampling(model, tokenizer, test_prompt, prob=0.3)
print(test_prompt, next_token, possible_tokens)
assert next_token in possible_tokens
assert sorted(possible_tokens) == ['identify', 'learn', 'predict', 'read', 'recognize', 'understand']
assert len(possible_tokens) == 6

In [ ]:
def get_topk_next_tokens(model, input_ids, beam_size):
  with torch.no_grad():
    logits = model(input_ids=input_ids).logits[0, -1, :]
    log_probs = torch.nn.functional.log_softmax(logits, dim=-1)

  top_k_score, top_k_ids = torch.topk(log_probs, beam_size)
  return top_k_score, top_k_ids

def expand_candidate(candidate, top_k_score, top_k_ids):
    candidates = []

    for i in range(len(top_k_ids)):
      new_candidate = {
          "tokens_ids": torch.cat([
              candidate['tokens_ids'],
              top_k_ids[i].view(1,1) # top_k_ids - [beam_size]
          ], dim=-1), # <- склеиваем по последней оси
          "score": candidate["score"] + top_k_score[i],
          "finished": top_k_ids[i].item() == tokenizer.eos_token_id
      }
      candidates.append(new_candidate)
    return candidates


In [ ]:
def beam_search(model, tokenizer, prompt: str, beam_size: int = 2, max_new_tokens: int = 10):
    """generates the next token from the beam search sampling"""

    inputs = tokenizer(prompt, return_tensors='pt', return_token_type_ids=False).to(device)
    top_k_score, top_k_ids = get_topk_next_tokens(model, inputs['input_ids'], beam_size)

    start_candidate = {
        "tokens_ids": inputs['input_ids'], # [batch_size, seq_len]
        "score": torch.tensor(0.0, device=device),
        "finished": False
    }

    # шаги генерации токенов
    beam = [start_candidate]
    for step in range(max_new_tokens):
        all_candidates = []

        for candidate in beam:
            if candidate['finished']:
                all_candidates.append(candidate)
                continue

            input_ids = candidate['tokens_ids']

            top_k_score, top_k_ids = get_topk_next_tokens(model, input_ids, beam_size)
            all_candidates += expand_candidate(candidate, top_k_score, top_k_ids)

        beam = sorted(
            all_candidates,
            key=lambda x: x['score'].item(),
            reverse=True
        )[:beam_size]

        if all(b['finished'] for b in beam):
            break
    best = beam[0]
    return tokenizer.decode(best['tokens_ids'][0], skip_special_tokens=True).strip()

In [ ]:
# Tests for nucleus sampling
test_prompt = "Elbrus is the highest"
candidates = beam_search(model, tokenizer, test_prompt, 3, 20)
print(candidates)

Elbrus is the highest mountain in the Russian Federation. It is located on the border between the European and Asian parts of Russia


### Part 3: Chain-of-thought prompting (4 points total)

![img](https://github.com/kojima-takeshi188/zero_shot_cot/raw/main/img/image_stepbystep.png)

---



In [ ]:
import json
import random
import locale; locale.getpreferredencoding = lambda: "UTF-8"
!wget https://raw.githubusercontent.com/kojima-takeshi188/zero_shot_cot/2824685e25809779dbd36900a69825068e9f51ef/dataset/AQuA/test.json -O aqua.json
data = list(map(json.loads, open("aqua.json")))

--2026-06-18 12:18:38--  https://raw.githubusercontent.com/kojima-takeshi188/zero_shot_cot/2824685e25809779dbd36900a69825068e9f51ef/dataset/AQuA/test.json
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133, 185.199.111.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 130192 (127K) [text/plain]
Saving to: ‘aqua.json’

aqua.json           100%[===================>] 127.14K  --.-KB/s    in 0.02s   

2026-06-18 12:18:38 (5.51 MB/s) - ‘aqua.json’ saved [130192/130192]



In [ ]:
len(data)

254

In [ ]:
print("Example:")
data[150]

Example:


{'question': 'Janice bikes at 10 miles per hour, while Jennie bikes at 20. How long until they have collectively biked 1 mile?',
 'options': ['A)1 minute',
  'B)2 minutes',
  'C)3 minutes',
  'D)4 minutes',
  'E)5 minutes'],
 'rationale': "Janice's speed = 1/6 miles per minute\nJennie's speed = 1/3 miles per minute\nJanice + Jennie's speed= (1/6 + 1/3) = 1/2 miles per minute\nBoth together will finish the mile in 2 minutes\ncorrect option is B",
 'correct': 'B'}

### Naive solution

Here, we prompt the model to choose an answer to the example above (`data[150]`) out of the options given above. We're using a format that mimics grade school solution textbook.

Please note that there are minor formatting changes in options: an extra space and an opening bracket. Those may or may not be important :)

In [ ]:
EXAMPLE_0SHOT = """
Question: Janice bikes at 10 miles per hour, while Jennie bikes at 20. How long until they have collectively biked 1 mile?
Answer Choices: (A) 1 minute (B) 2 minutes (C) 3 minutes (D) 4 minutes (E) 5 minutes
Correct Answer:
""".strip()

In [ ]:
# solving an equation directly
batch = tokenizer(EXAMPLE_0SHOT, return_tensors='pt', return_token_type_ids=False).to(device)
torch.manual_seed(1337)
output_tokens = model.generate(**batch, max_new_tokens=100, do_sample=True, top_p=0.9)
print("[Prompt:]\n" + EXAMPLE_0SHOT)
print("=" * 80)
print("[Generated:]", tokenizer.decode(output_tokens[0][batch['input_ids'].shape[1]:].cpu()))

[Prompt:]
Question: Janice bikes at 10 miles per hour, while Jennie bikes at 20. How long until they have collectively biked 1 mile?
Answer Choices: (A) 1 minute (B) 2 minutes (C) 3 minutes (D) 4 minutes (E) 5 minutes
Correct Answer:
[Generated:]  D
We can solve this problem by setting up a ratio:
Jennie bikes for 1/2 the time that Janice does.
If we let x be the amount of time it takes for both to bike 1 mile individually,
then it will take 2x for them to bike 2 miles together and 4x for them to bike 4 miles together.
Therefore, x = .06 hours or 3.6 minutes.
The answer is (C).<|endoftext|>


And here's how you can solve this with few-shot chain-of-thought prompting.

You need to chang 3 things
- use a new field called **Rationale**, that contains a step-by-step solution to the problem
- add several few-shot examples of previously solved problems **with rationales**
- change the final prompt so that the model has to generate rationale before answering

In [ ]:
EXAMPLE_3SHOT_CHAIN_OF_THOUGHT = """
Question: The original retail price of an appliance was 60 percent more than its wholesale cost. If the appliance was actually sold for 20 percent less than the original retail price, then it was sold for what percent more than its wholesale cost?
Answer Choices: (A) 20% (B) 28% (C) 36% (D) 40% (E) 42%
Rationale: wholesale cost = 100;\noriginal price = 100*1.6 = 160;\nactual price = 160*0.8 = 128.\nAnswer: B.
Correct Answer: B


Question: A grocer makes a 25% profit on the selling price for each bag of flour it sells. If he sells each bag for $100 and makes $3,000 in profit, how many bags did he sell?
Answer Choices: (A) 12 (B) 16 (C) 24 (D) 30 (E) 40
Rationale: Profit on one bag: 100*1.25= 125\nNumber of bags sold = 3000/125 = 24\nAnswer is C.
Correct Answer: C


Question: 20 marbles were pulled out of a bag of only white marbles, painted black, and then put back in. Then, another 20 marbles were pulled out, of which 1 was black, after which they were all returned to the bag. If the percentage of black marbles pulled out the second time represents their percentage in the bag, how many marbles in total Q does the bag currently hold?
Answer Choices: (A) 40 (B) 200 (C) 380 (D) 400 (E) 3200
Rationale: We know that there are 20 black marbles in the bag and this number represent 1/20 th of the number of all marbles in the bag, thus there are total Q of 20*20=400 marbles.\nAnswer: D.
Correct Answer: D


Question: Janice bikes at 10 miles per hour, while Jennie bikes at 20. How long until they have collectively biked 1 mile?
Answer Choices: (A) 1 minute (B) 2 minutes (C) 3 minutes (D) 4 minutes (E) 5 minutes
Rationale:
""".strip()

In [ ]:
batch = tokenizer(EXAMPLE_3SHOT_CHAIN_OF_THOUGHT, return_tensors='pt', return_token_type_ids=False).to(device)
torch.manual_seed(1337)
output_tokens = model.generate(**batch, max_new_tokens=100, do_sample=True, top_p=0.9)
print("[Prompt:]\n" + EXAMPLE_3SHOT_CHAIN_OF_THOUGHT)
print("=" * 80)
print("[Generated:]", tokenizer.decode(output_tokens[0][batch['input_ids'].shape[1]:].cpu()))
#### NOTE: scroll down for the final answer (below the ======= line)

[Prompt:]
Question: The original retail price of an appliance was 60 percent more than its wholesale cost. If the appliance was actually sold for 20 percent less than the original retail price, then it was sold for what percent more than its wholesale cost?
Answer Choices: (A) 20% (B) 28% (C) 36% (D) 40% (E) 42%
Rationale: wholesale cost = 100;
original price = 100*1.6 = 160;
actual price = 160*0.8 = 128.
Answer: B.
Correct Answer: B


Question: A grocer makes a 25% profit on the selling price for each bag of flour it sells. If he sells each bag for $100 and makes $3,000 in profit, how many bags did he sell?
Answer Choices: (A) 12 (B) 16 (C) 24 (D) 30 (E) 40
Rationale: Profit on one bag: 100*1.25= 125
Number of bags sold = 3000/125 = 24
Answer is C.
Correct Answer: C


Question: 20 marbles were pulled out of a bag of only white marbles, painted black, and then put back in. Then, another 20 marbles were pulled out, of which 1 was black, after which they were all returned to the bag. If 

__Task 6 (1 pt)__ write a function that automatically creates chain-of-thought prompts. Follow the instructions from the function docstring.

Example:
{'question': 'Janice bikes at 10 miles per hour, while Jennie bikes at 20. How long until they have collectively biked 1 mile?',
 'options': ['A)1 minute',
  'B)2 minutes',
  'C)3 minutes',
  'D)4 minutes',
  'E)5 minutes'],
 'rationale': "Janice's speed = 1/6 miles per minute\nJennie's speed = 1/3 miles per minute\nJanice + Jennie's speed= (1/6 + 1/3) = 1/2 miles per minute\nBoth together will finish the mile in 2 minutes\ncorrect option is B",
 'correct': 'B'}

In [ ]:
QUESTION_PREFIX = "Question: "
OPTIONS_PREFIX = "Answer Choices: "
CHAIN_OF_THOUGHT_PREFIX = "Rationale: "
ANSWER_PREFIX = "Correct Answer: "
FEWSHOT_SEPARATOR = "\n\n\n"

def make_prompt(*, main_question, fewshot_examples):
  """
  Your goal is to produce the same prompt as the EXAMPLE_3SHOT_CHAIN_OF_THOUGHT automatically

  For each few-shot question, make sure to follow the following rules:
  1. Each question begins with QUESTION_PREFIX, after which you should print the question without leading/traiiling spaces (if any)
  2. After the question, provide space-separated options. Each option should be put in double brackets, followed by option text, e.g. "(A) 146%"
  3. Then, provide the answer as a single letter (A-E)
  4. Finally, add trailing newlines from FEWSHOT_SEPARATOR

  Your final prompt should contain all fewshot_examples (in order), separated with FEWSHOT_SEPARATOR, then follow with main_question.
  The main_question should contain the question and options formatted the same way as in FEWSHOT_EXAMPLES.
  After that, you should prompt the model to produce an explanation (rationale) for the answer.

  Please make sure your prompt contains no leading/trailing newlines or spaces, same as in EXAMPLE_3SHOT_CHAIN_OF_THOUGHT
  """
  def format_options(options):
      return " ".join("(" + option.replace(")", ") ") for option in options)

  def format_example(example, with_answer=True):
      parts = [
          QUESTION_PREFIX + example["question"].strip(),
          OPTIONS_PREFIX + format_options(example["options"]),
          CHAIN_OF_THOUGHT_PREFIX + example["rationale"].strip(),
      ]

      if with_answer:
          parts.append(ANSWER_PREFIX + example["correct"].strip())

      return "\n".join(parts)

  fewshot_part = FEWSHOT_SEPARATOR.join(
      format_example(example, with_answer=True)
      for example in fewshot_examples
  )

  main_part = "\n".join([
      QUESTION_PREFIX + main_question["question"].strip(),
      OPTIONS_PREFIX + format_options(main_question["options"]),
      CHAIN_OF_THOUGHT_PREFIX,
  ])

  if fewshot_part:
      return (fewshot_part + FEWSHOT_SEPARATOR + main_part).strip()

  return main_part.strip()



generated_fewshot_prompt = make_prompt(main_question=data[150], fewshot_examples=(data[30], data[20], data[5]))
assert generated_fewshot_prompt == EXAMPLE_3SHOT_CHAIN_OF_THOUGHT, "prompts don't match"
assert generated_fewshot_prompt != make_prompt(main_question=data[150], fewshot_examples=())
assert generated_fewshot_prompt.endswith(make_prompt(main_question=data[150], fewshot_examples=()))

print("Well done!")

# Hint: if two prompts do not match, you may find it usefull to use https://www.diffchecker.com or similar to find the difference

Well done!


__Task 7 (1 points):__ Evaluate your prompt.

Please run the model on the entire dataset and measure it's accuracy.
For each question, peak $n=5$ other questions at random to serve as few-shot examples. Make sure not to accidentally sample the main_question among few-shot examples. For scientific evaluation, it is also a good practice to split the data into two parts: one for eval, and another for few-shot examples. However, doing so is optional in this homework.

The tricky part is when to stop generating: if you don't control for this, your model can accidentally generate a whole new question - and promptyly answer it :) To make sure you get the correct answer, stop generating tokens when the model is done explaining it's solution. To circumvent this, you need to __stop generating as soon as the model generates Final Answer: [A-E]__
To do so, you can either generate manually (see low-level generation above) or use [transformers stopping criteria](https://discuss.huggingface.co/t/implimentation-of-stopping-criteria-list/20040/2), whichever you prefer.

If you do everything right, the model should be much better than random. However, please __do not expect miracles__: this is far from the best models, and it will perform much worse than an average human.

In [ ]:
NUM_SAMPLES = 0    # use this to count how many samples you evaluated
NUM_RESPONDED = 0  # how many times did the model produce Correct Answer: (letter) in it's response. use as a sanity check.
NUM_CORRECT = 0    # how many times did the model's chosen answer (letter) match the correct answer

In [ ]:
import re
import numpy as np
from datetime import datetime


exp = []

for i, item in enumerate(data[:50]):
    start_time = datetime.now()

    fewshot_indices = random.sample(range(len(data) - 1), 5)
    fewshot_indices = [j if j < i else j + 1 for j in fewshot_indices]
    few_shot = [data[j] for j in fewshot_indices]


    prompt = make_prompt(main_question=item, fewshot_examples=few_shot)
    batch = tokenizer(prompt, return_tensors='pt', return_token_type_ids=False).to(device)
    output_tokens = model.generate(**batch, max_new_tokens=200, do_sample=False)
    output_text = tokenizer.decode(output_tokens[0][batch['input_ids'].shape[1]:].cpu(), skip_special_tokens=True)


    match = re.search(r"(?:Final\s+Answer|Correct\s+Answer|Answer)\s*:?\s*(?:Option\s*)?([A-E])\b", output_text)

    if match:
        answer = match.group(1)
    else:
        answer = None

    NUM_SAMPLES += 1
    end_time = datetime.now()

    exp_time = (end_time - start_time).total_seconds()
    exp.append(exp_time)

    if NUM_SAMPLES % 10 == 0:
      print(f"avg time: {sum(exp) / len(exp)}")

    if answer is not None:
        NUM_RESPONDED += 1
        if item["correct"] == answer:
            NUM_CORRECT += 1


print(f"all avg time: {sum(exp) / len(exp)}")

avg time: 8.461498666666666
avg time: 7.8730831052631585
avg time: 7.870075689655172
avg time: 7.877361179487179
avg time: 7.907134306122449
all avg time: 7.926221839999999


In [ ]:
print("Responded %%:", NUM_RESPONDED / NUM_SAMPLES)
print("Accuracy (when responded):", NUM_CORRECT / NUM_RESPONDED)
print("Accuracy (overall):", NUM_CORRECT / NUM_SAMPLES)

if NUM_RESPONDED / NUM_SAMPLES < 0.9:
  print("Something is wrong with the evaluation technique (for 5-shot CoT): the model refuses to answer too many questions.")
  print("Make sure you generate enough tokens that the model can produce a correct answer.")
  print("When in doubt, take a look at the full model output. You can often spot errors there.")

Responded %%: 0.42857142857142855
Accuracy (when responded): 0.6153846153846154
Accuracy (overall): 0.26373626373626374
Something is wrong with the evaluation technique (for 5-shot CoT): the model refuses to answer too many questions.
Make sure you generate enough tokens that the model can produce a correct answer.
When in doubt, take a look at the full model output. You can often spot errors there.
